### DISCUSSION TITLE
Use code instead of the packages

### DISCUSSION THREAD

**Discussion by riatdyl** [elsa-workflows/elsa-core#906] [Q&A] [answered]
Hello,

I would like to use the code instead of the packages, is it possible ?

This request is because I want to use my own activities and remove the existing ones.
And btw I would edit the activities in the packages.

I tried to remove the "usings" from "Startup.cs" and the ".AddHttpActivities", ".AddEmailActivities", etc. but the activities are always there in the Designer.

**Comment by sfmskywalker** [comment, marked as answer]
That is certainly possible. You could fork the project and make any changes you want.

However, it might be more interesting still to use the packages, since they would include new features and bug fixes. You can still get those by updating your fork of course, but might be less convenient.

When using the packages, you can still add your own activities without changing Elsa itself.

Regarding the ability to not use any of the existing activities, you can make that work too. All you would have to do from your own startup class is call this:

```csharp
services.AddElsa(elda => elsa.NoCoreActivities())
```

PS. This is all Elsa 2. For Elsa 1, you would have to fork the project.

**Reply by riatdyl** [reply]
Okay, I tried to do this but I have an error.

This is the working code :
```c#
public void ConfigureServices(IServiceCollection services)
        {
            services
                // Add services used for the workflows runtime.
                .AddElsa(elsa => elsa.AddEntityFrameworkStores<SqlServerContext>(options => options.UseSqlServer(Configuration.GetConnectionString("SqlServer"))))
                .AddHttpActivities(options => options.Bind(Configuration.GetSection("Elsa:Http")))
                .AddEmailActivities(options => options.Bind(Configuration.GetSection("Elsa:Smtp")))
                .AddTimerActivities(options => options.Bind(Configuration.GetSection("Elsa:Timers")))

                // Add services used for the workflows dashboard.
                .AddElsaDashboard();
        }
```
\
And this is what I tried to do with this error : 
**'ElsaBuilder' does not contain a definition for 'NoCoreActivities' and no accessible extension method 'NoCoreActivities' accepting a first argument of type 'ElsaBuilder' could be found (are you missing a using directive or an assembly reference?)**
```c#
public void ConfigureServices(IServiceCollection services)
        {
            services
                // Add services used for the workflows runtime.
                .AddElsa(elsa => elsa.NoCoreActivities())

                // Add services used for the workflows dashboard.
                .AddElsaDashboard();
        }
```

**Reply by sfmskywalker** [reply]
Like I said, it only works for Elsa 2. You are using Elsa 1.

**Reply by riatdyl** [reply]
Oh sorry, I didn't know it was Elsa 1.

I used the Elsa Dashboard Guide and I thought it was the last version.

How can I use the Elsa 2 ?

**Reply by sfmskywalker** [reply]
No problem. I should have clarified that Elsa 2 is currently only available as a preview release, not as the "latest" release. We haven't even got documentation for it yet.

If you want to try Elsa 2, you should install the `Elsa` package from the [MyGet feed](https://www.myget.org/gallery/elsa-2) (make sure to select "Preview" or "Prelease" option from your NuGet package manager) and take a look at some of the sample projects you can find [here](https://github.com/elsa-workflows/elsa-core/tree/feature/elsa-2.0/src/samples).

We'll be working on guides & docs hopefully next month.

**Reply by riatdyl** [reply]
Thank you a lot, I'll check that.

Do you know when your project will be completed ?

I support you for your efforts, keep it up !

**Reply by sfmskywalker** [reply]
The plan is to mark the current version as a Release Candidate release at the end of this week. Which means that with luck, we can call it an official 2.0 release a couple weeks after. This doesn't mean it will be a completed project of course; there are a ton of additional features and enhancements we have in mind :)

**Reply by riatdyl** [reply]
> If you want to try Elsa 2, you should install the `Elsa` package from the [MyGet feed](https://www.myget.org/gallery/elsa-2) (make sure to select "Preview" or "Prelease" option from your NuGet package manager) and take a look at some of the sample projects you can find [here](https://github.com/elsa-workflows/elsa-core/tree/feature/elsa-2.0/src/samples).

I tried to launch one of the samples (src/samples/dashboard/aspnetcore/ElsaDashboard.Samples.AspNetCore.Monolith) and I have an error : **ERR_CONNECTION_RESET**

Do you have an idea on how to solve it ?

**Reply by sfmskywalker** [reply]
Sure. So, since the samples are (currently) provided as part of the Elsa source code, you will have to build the client-side (JS, CSS etc) assets yourself first (you need NodeJS installed for this to work). The AppVeyor build pipeline does this upon build, but you can do it yourself (make sure to use the correct paths - I'll just specify relative paths below):

```
cd .\src\designer\elsa-workflows-studio
npm install
npm run build

cd ..\bindings\aspnet\Elsa.Designer.Components.Web
npm install
npm run build

```

If you prefer not to deal with this, you can always recreate the sample projects yourself and just reference the Elsa packages instead of the source code projects.

**Reply by riatdyl** [reply]
Okay, I will try to do this.

What is this AppVeyor ?

And where are the references you're talking about ?

Sorry for all these questions, I'm a bit lost ^^'

**Reply by sfmskywalker** [reply]
No worries :)

AppVeyor: https://www.appveyor.com/
NodeJS: https://nodejs.org/en/about/

We use AppVeyor to automatically produce NuGet packages whenever we push to the main branch or Elsa 2 branch.
NodeJS is used to run NPM scripts to produce client-side assets.

I remember well when I first encountered these things - I was completely lost myself 😅

**Reply by riatdyl** [reply]
> cd .\src\designer\elsa-workflows-studio
> npm install
> npm run build
> 
> cd ..\bindings\aspnet\Elsa.Designer.Components.Web
> npm install
> npm run build
> ```

I executed theses commands in the right folders, but I always have the error I mentioned :(

**Reply by giannik** [reply]
Hello,
I tried the steps above  and got the following message (missing .css file) when running ElsaDashboad.Samples.Monolith.


Oh I also need to run npm on the  following to get it up to here:

cd ..src\dashboards\blazor\ElsaDashboard.Application
npm install
npm run build


![image](https://user-images.githubusercontent.com/1516981/116773265-29820780-aa5d-11eb-85fc-aa9cab1694a8.png)

**Reply by riatdyl** [reply]
Hello,

I tried to run npm install and npm run build where you said.

The problem is that the program executes with no error, but I always have the ERR_CONNECTION_RESET.

I tried to launch ElsaDashboard.Samples.AspNetCore.Monolith and ElsaDashboard.Samples.Monolith but there's the same error.

**Reply by riatdyl** [reply]
> Sure. So, since the samples are (currently) provided as part of the Elsa source code, you will have to build the client-side (JS, CSS etc) assets yourself first (you need NodeJS installed for this to work).

So, where do I have to create the client side assets ?

**Reply by riatdyl** [reply]
Okay I tried to launch the application normally without IIS Express (default) and it works now !

**Reply by sfmskywalker** [reply]
Just in case anyone else runs into this issue: to ensure static assets are served even when running in IIS Express and/or in an environment that is not `Development` and the project is not published, apply the following in Program.cs:

`.UseStaticWebAssets()`

Like so:

```csharp
Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder(args)
   .ConfigureWebHostDefaults(webBuilder => webBuilder
      .UseStaticWebAssets()
      .UseStartup<Startup>())
```

See also [this documentation](https://docs.microsoft.com/en-us/aspnet/core/razor-pages/ui-class?view=aspnetcore-5.0&tabs=visual-studio#consume-content-from-a-referenced-rcl-1).
